lisa/wlgen/workload.py

```diff
@@ -170,17 +203,13 @@ def run(self, cpus=None, cgroup=None, background=False, as_root=False, timeout=N
         command = self.wrap_command(cpus, cgroup)
 
         logger.info("Execution start: %s", command)
+        self.output = target.execute(command, as_root=as_root, timeout=timeout)
```

Comment:
Wouldn't it make more sense for `run()` to return the output, rather than storing it as an attribute ? Avoiding statefull APIs usually reduce the confusion. With state like that, it's not always obvious if a given workload is suitable or not: does it have an output yet or do we need to run it first ? Then the client code will add tests to check if that attribute is None or not, and `__init__` will initialize that to `None`.  We can avoid all this confusion and boilerplate by just having objects with a single responsibility: representing the workload, and another one representing the result of an execution. These two things are separate, and if merged together usually give rise to half of the methods checking some attributes for `None` and raising `RuntimeError` at the beginning of the method (like we used to have in `Trace` for example).

Reply:
I think it makes sense, I just need to have a look around and see how many users of this API there is.